py/objtuple.c

```diff
@@ -249,7 +252,7 @@ mp_obj_t mp_obj_new_tuple(size_t n, const mp_obj_t *items) {
 }
 
 void mp_obj_tuple_get(mp_obj_t self_in, size_t *len, mp_obj_t **items) {
-    assert(mp_obj_is_type(self_in, &mp_type_tuple));
+    mp_check_self(mp_obj_is_tuple_compatible(self_in));
```

Comment:
As you found out, `mp_obj_tuple_get` is really an (internal) C-API function.  So it is assumed that the caller knows what they are doing, and therefore there's an `assert` rather than an `mp_check_self`.  The latter is only used when incoming objects might be of the wrong type due to a user's script.

So this should really still be an `assert`.  Then the change in this PR just opens up the possibility that this C-API function can have tuple-compatible objects passed in.

Reply:
Done.  I also renamed the macro _mp_obj_is_tuple_compatible so future readers of the code will suspect it does not follow the usual pattern of mp_obj_is_TYPE.

Reply:
I also changed the other site from using mp_check_self to assert, as it appears to be "sufficiently internal" as well.